### ISSUE TITLE
Customization of the juice shop

### ISSUE THREAD

**Issue by wurstbrot** [juice-shop/juice-shop#277]
As a security awareness trainer in a company I want to shock and fear my audience as much as possible.
Therefore I need to customize the juice shop with the look and feel of the company I am giving the training to.

The following things need to be done:
- [x] - Changing the name and title of the shop
- [x] - Replacement of all occurrences of juice-sh.op with a configurable domain (so you can login for example with admin@<domain>
- [x] - Adding of x random users
- [x] - Creation of customized products to create a relation for the audience (make sure "Change Product Challenge" and "Christmas Special Challenge" is still working)
- [x] - Do not display notifications of solved challenges as they destroy the flow of a story
- [x] - Make bootstrap theme configurable for be able to pick one close to the corporate color scheme
- [x] Make the Twitter and Facebook URLs on the /basket screen configurable
- [x] Make the `favicon.ico` customizable (or at least removable if set to null)

I created a draft shell script which does everything described except deletion of existing products:
https://github.com/wurstbrot/juice-shop/blob/master/customize.bash

In case the script suits the need for contributing, I will create a pull request. Otherwise I have the following suggestions to implement it:
Configuration takes place over a json file with default values:
```json
{
	"domain": "juice-sh.op",
	"companyName": "Juice Shop",
	"logoUrl": "",
	"numberOfFakeUsers": 100,
	"challengeNotifications": true,
	"products": [{
		"name": "Apple Juice (1000ml)",
		"description": "The all-time classic.",
		"price": 1.99,
		"image": "apple_juice.jpg"
	}, {
		"name": "Orange Juice (1000ml)",
		"description": "Made from oranges hand-picked by Uncle Dittmeyer.",
		"price": 2.99,
		"image": "orange_juice.jpg"
	}, {
		"name": "Eggfruit Juice (500ml)",
		"description": "Now with even more exotic flavour.",
		"price": 8.99,
		"image": "eggfruit_juice.jpg"
	}]
}
```
The company name will be used from the configuration file, e.g. for title. 
If no logoUrl is given, it will use the default file. Otherwise on startup the logo will be downloaded and replaced on the file system. This allows the usage of the shop without internet connection.

To integrate users and products, the function createProduct in data/datacreator.js will iterate through the defined products. In case that a needed object name like image is not given, it needs to use a default one. In case the price or description is not given, a random price/name will be generated. Images for products will be downloaded on startup.

The standard users will be created as before with the given domain. Every file which contains the domain needs to use the configuration domain (e.g. tests). Extra random users will be generated with random emails and passwords.

Notifications can be hidden by adding `.alert {display:none;}` to app/css/app.css

The suggestions here are to discus how I will implement it.

**Comment by bkimminich** [comment]
The idea in general I like a lot! Here are some spontaneous considerations without having tested the script yet:

* Hiding all `.alert` would also hide all regular error messages from the UI, making some vulnerabilities harder to explain and reducing the "happy path" UX
* I wouldn't use any real domains for the fake users to prevent generating an existing one by accident
* If you delete the existing products, some challenges probably won't work any more. I get that this might be acceptable for awareness demos, where you don't plan to show the Score Board at all. But I think the redressing would also be great for corporate trainings, so a "challenge-preserving" script would be preferable. But that could come in different steps obviously.
* Having a `json` or `yml` configuration file would be wonderful, especially if it worked on all OS then. But to begin with a Shell script with in place config is totally fine.
* For squelching the challenge notifications these could get their own CSS class.

**Comment by wurstbrot** [comment]
Thank you for your feedback.

I tried to use angulars ng-bind attribute to set the title/name in index.html. You can see my (not working) implementation in [app.js and index.html](https://github.com/wurstbrot/juice-shop/commit/cc46c0c0b9a4dea4d45e8d26ab7321c51da82be3). Do you have any idea why it is not working like for example isLoggedIn()?

**Comment by bkimminich** [comment]
Hm, I implemented it similarly (see bc73dad) and changing the title works for me. Did you `npm install` (or at least `grunt minify`) before you tested?

**Comment by bkimminich** [comment]
Hm, I implemented it similarly (see bc73dad and changing the title works for me. Did you `npm install` (or at least `grunt minify`) before you tested?

**Comment by wurstbrot** [comment]
You got me.

**Comment by bkimminich** [comment]
@wurstbrot: A little _extrawurst_ that might be nice: Making the Twitter and Facebook URLs on the `/basket` screen configurable?

**Comment by bkimminich** [comment]
🐛 ... The app crashes on Heroku immediately after start because it cannot bind the port for some reason... I tried to fix that with 6682c8b53efe6f8b39804cac8635d0be8cf32693, let's see if it helps...
```
2017-03-26T22:28:31.378283+00:00 heroku[web.1]: Error R10 (Boot timeout) -> Web process failed to bind to $PORT within 60 seconds of launch
2017-03-26T22:28:31.378433+00:00 heroku[web.1]: Stopping process with SIGKILL
2017-03-26T22:28:31.484523+00:00 heroku[web.1]: Process exited with status 137
2017-03-26T22:28:31.500298+00:00 heroku[web.1]: State changed from starting to crashed
2017-03-26T22:28:33.929975+00:00 heroku[router]: at=error code=H10 desc="App crashed" method=GET path="/" host=juice-shop-staging.herokuapp.com request_id=9b45a8ed-336f-4cd3-b0d4-05a63fe2a6a4 fwd="77.187.235.155" dyno= connect= service= status=503 bytes= protocol=https
2017-03-26T22:28:34.626562+00:00 heroku[router]: at=error code=H10 desc="App crashed" method=GET path="/favicon.ico" host=juice-shop-staging.herokuapp.com request_id=5f055243-d919-46e7-bfa8-bc8880acbf98 fwd="77.187.235.155" dyno= connect= service= status=503 bytes= protocol=https
2017-03-26T22:28:39.406847+00:00 heroku[router]: at=error code=H10 desc="App crashed" method=GET path="/" host=juice-shop-staging.herokuapp.com request_id=5d2b1ef9-4ae8-4a79-bc24-a7f46570d86b fwd="77.187.235.155" dyno= connect= service= status=503 bytes= protocol=http
2017-03-26T22:28:39.941948+00:00 heroku[router]: at=error code=H10 desc="App crashed" method=GET path="/favicon.ico" host=juice-shop-staging.herokuapp.com request_id=4ccc6ece-f35c-4314-9088-7f40c7b22444 fwd="77.187.235.155" dyno= connect= service= status=503 bytes= protocol=http
```

**Comment by bkimminich** [comment]
~And another immersion extra we might add: Either making the pre-defined Feedbacks completely neutral (e.g. referring to "site" instead of "shop" or anything) or using the app name in them. Additionally the slideshow images could be customizable...~

Feedback "neutralization" implemented with	b8b441d.

**Comment by bkimminich** [comment]
I just removed the subtask to change `{{juiceshop}}` from OWASP Juice Shop into the customized application name. This is only used in the request for donation and contribution to the project itself, so having the actual project name in there is correct.
